### PR TITLE
Fix honoring tmpfs-size for user /dev/shm mount

### PR DESCRIPTION
**Pull request by kolyshkin**
**- What I did**
Commit 7120976d74195 ("Implement none, private, and shareable ipc modes") introduces a bug: if a user-specified mount for /dev/shm is provided, its size is overridden by value of ShmSize. This was reported in https://github.com/moby/moby/issues/35271.

This commit is an attempt to fix this, as well as optimize things a but and make the code easier to read.

**- How I did it**
Appending the 'size' argument to /dev/shm mount options moved to an earlier place in the function, where we deal with the mounts that came from spec, therefore a user mount is left intact.

**- How to verify it**

There is a test case included in this PR. In addition, it can easily be checked manually:
```
 docker run --rm \
	--mount type=tmpfs,dst=/dev/shm,tmpfs-size=100K \
	alpine df /dev/shm
```

Expected output (i.e. if there is no bug):
```
Filesystem           1K-blocks      Used Available Use% Mounted on
tmpfs                      100         0       100   0% /dev/shm
```

(i.e. `100` should be shown under the "1K-blocks" header)

**- Description for the changelog**
Fix honoring tmpfs-size for user /dev/shm mount

**- A picture of a cute animal (not mandatory but encouraged)**
shy-beaver
*Shy beaver | Minnesota zoo ([source](https://www.flickr.com/photos/nouqraz/198455748))*

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
could you add a test?

**Comment by kolyshkin** [comment]
> could you add a test?

What stops me is lack of `client.ContainerExec` functionality in `integration/` tests, something similar to `cli.DockerCmd("exec", ...)` in `integration-cli/`. Guess I need to write it first.

**Review by dnephin — COMMENTED** [review]
Would it be possible to test this with a unit test instead of an integration that?  The fix seems to be in a single function, so having to make multiple API calls seems like it's maybe not the easiest way to reproduce and test the issue.

**Comment by kolyshkin** [comment]
@AkihiroSuda @thaJeztah I have added the test case, and made sure it is failing without the fix (and passing with the fix applied).

**Comment by kolyshkin** [comment]
> Would it be possible to test this with a unit test instead of an integration that?

Problem is, there's a lot of code dealing with mounts, and the `setMounts()` is not very isolated function. If I am to write a unit test, I have to repeat some of that code in the test case which is prone to error later if this code changes.

Anyway, let me give it a try and see what I end up with.

**Comment by kolyshkin** [comment]
@dnephin OK, I managed to write a unit test. PTAL once time permits.

File: daemon/daemon_linux_test.go [hunk 1]
```diff
@@ -102,3 +109,49 @@ func TestNotCleanupMounts(t *testing.T) {
...
+				if !strings.HasPrefix(o, "size=") {
+					continue
+				}
+				t.Logf("%+v\n", m.Options)
+				assert.Equal(t, "size="+size, o)
+				found = true
+			}
+		}
+	}
+	assert.Equal(t, true, found)
```
Comment (AkihiroSuda) [inline daemon/daemon_linux_test.go]: you can use assert.Truef
https://godoc.org/github.com/stretchr/testify/assert#Truef

Comment (dnephin) [reply daemon/daemon_linux_test.go]: I think this is one of the few times when a plain `t.Fatal()` is actually the best option:

```go
t.Fatal("size option was missing from mount")
```

**Review by dnephin — APPROVED** [review]
Thanks, test LGTM, just one small nit

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

**Comment by thaJeztah** [comment]
Looks good, but haven't given it a spin yet; let me check if the `--tmpfs /dev/shm:rw,nosuid,nodev,size=8` option (more options other than just size) works as expected.

File: daemon/daemon_linux_test.go (hunk 1 above)
Comment (kolyshkin) [reply daemon/daemon_linux_test.go]: updated to use f.Fatal()

**Comment by vieux** [comment]
@thaJeztah let us know :D

**Review by thaJeztah — APPROVED** [review]
Doh! I see I mixed up your PR's, had the other PR (https://github.com/moby/moby/pull/35467) in mind when I commented.

this LGTM, thanks!

**Comment by thaJeztah** [comment]
Oh, looks like there's one linting failure:

```
22:04:28 daemon/daemon_linux_test.go:156:5:warning: should omit comparison to bool constant, can be simplified to !found (S1002) (gosimple)
22:04:28 Build step 'Execute shell' marked build as failure
```

**Pull request merged**